proposals/p2240.md

```diff
@@ -0,0 +1,758 @@
+
+```
+fn SumInts((params: [i64;])[:] ...,) -> i64
+```
+
+To state the point more generally: when an expression mixes prefix and suffix
+operators, their order is determined by precedence and parentheses, which is
+usually less readable than relying on the linear order of operators with the
+same fixity. The binding operator `:` is effectively a prefix operator (because
+its left-hand side is a single token), so if `[:]` is a suffix operator, code
+that mixes the two will be somewhat harder to read. That's a particular problem
+because, as our motivating examples illustrate, variadic function signatures
+will often have those two operators in close conjunction, and readability is at
+a premium in function signatures.
+
+C solved a very similar problem with mixing prefix `*` and postfix `.` by
+introducing a separate `->` operator. We could in principle solve this issue the
+same way, introducing a separate "variadic binding operator" `$`, such that
+`params$ [i64;]` is equivalent to `(params: [i64;])[:]`. However, this would
+trade a readability problem for a teachability problem, by introducing more
+syntax that programmers need to learn before they can even read variadic
+function signatures.
+
+### Keyword syntax
+
+> **TODO:** Decide whether to make this the primary proposal, and restructure
+> the discussion as appropriate.
+
+We could build the syntax around keywords rather than symbolic tokens, much like
+the
+[Swift proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md).
+For concreteness, the keywords I will discuss here are as follows:
+
+| Symbolic token | Keyword replacement               |
+| -------------- | --------------------------------- |
+| `[:]`          | `expand`                          |
+| `...,`         | `repeat` and `unpack` (see below) |
+| `...and`       | `all_of`                          |
+| `...or`        | `any_of`                          |
+| `...{`         | `repeat {`                        |
+
+This approach seems to work best in combination with some other nominally
+independent changes:
+
+-   Forbid `expand` in pattern contexts.
+-   Allow `repeat` only in expression contexts and `unpack` only in pattern
+    contexts.
+-   Require any usage of a pack-type variable to be prefixed with a new keyword
+    `each`.
+
+With these changes, the examples given earlier would read as follows:
+
+```carbon
+// Computes the sum of its arguments, which are i64s
+fn SumInts(unpack params: expand [i64;]) -> i64 {
+  var sum: i64 = 0;
+  repeat {
+    sum += each params;
+  }
+  return sum;
+}
+```
+
+```carbon
+// Concatenates its arguments, which are all convertible to String
+fn StrCat[T:! [ConvertibleToString;]](unpack params: expand T) -> String {
+  var len: i64 = 0;
+  repeat { len += (each params).Length(); }
+  var result: String = "";
+  result.Reserve(len);
+  repeat { result.Append((each params).ToString()); }
+  return result;
+}
```

Comment:
```suggestion
// Concatenates its arguments, which are all convertible to String
fn StrCat[...each T:! ConvertibleToString](...each param: each T) -> String {
  var len: i64 = 0;
  ...{ len += each param.Length(); }
  var result: String = "";
  result.Reserve(len);
  ...{ result.Append(each param.ToString()); }
  return result;
}
```

Reply:
I think the mental model we'd want to teach for this `each` keyword is that `each param` should be treated as the name of the parameter. If people (and syntax highlighers) treat it that way, I think the parentheses in `(each param).Length()` might look as redundant as `(x).Length()` does. But I don't know that; it might be that visually splitting things up on whitespace is just too ingrained a habit for lots of readers. The misinterpretation is probably also not too harmful for readers of code, other than letting them infer an inaccurate rule -- reading `each param.Length()` as `each (param.Length())` or reading `each x + 1` as `each (x + 1)` still gives the right interpretation -- and for writers of code we should be able to give great diagnostics if the `each` is misplaced (eg, `each 1 - x` instead of `1 - each x`).